File: NoughtsAndCrosses/Connection/HTTP/HttpSessionBasedOnJson.cs
using System;
using System.Collections.Generic;
using Newtonsoft.Json.Linq;
using NoughtsAndCrosses.Utils;

namespace NoughtsAndCrosses.Connection.HTTP {
  public class HttpSessionBasedOnJson : HttpSession {

    public override void ReceiveData(string command, byte[] data, int size) {
      if (!inMessageHandlers.ContainsKey(command)) {
        OnReceivingError(RECEIVE_FATAL_ERROR, "Internal error: unknown command");
        return;
      }
      try {
        DataReader dataReader = new DataReader(data, 0, size);
        string json = String.Empty;
        dataReader.ReadString(ref json);
        var jsonObject = JObject.Parse(json);
        string status = (string)jsonObject["status"];
        if (status == "error") {
          JToken token = null;
          if (jsonObject.TryGetValue("msg_id", out token)) {
            OnReceivingError(command, (int)token, (string)jsonObject["message"]);
            return;
          }
          OnReceivingError(command, (string)jsonObject["message"]);
          return;
        }
        if (status != "ok") {
          OnReceivingError(command, "Unknown error");
          return;
        }
        // Обрабатываем сообщение - вызываем соответствующий обработчик
        inMessageHandlers[command](jsonObject);
      }
      catch (Exception exc) {
        string s = string.Format("Internal error {0}", exc.Message);
        OnReceivingError(Session.RECEIVE_FATAL_ERROR, s);
        return;
      }
    }

    public HttpSessionBasedOnJson(IConnectManager connManager, IConnectionInfo connection)
    : base(connManager, connection) {

    }

    protected delegate bool JSONMessageHandler(JObject jsonObject);

    private Dictionary<object, JSONMessageHandler> inMessageHandlers = new Dictionary<object, JSONMessageHandler>();

    protected void addMessageHandler(object type, JSONMessageHandler messageHandlers) {
      inMessageHandlers.Add(type, messageHandlers);
    }

    protected virtual void OnReceivingError(string type, int msgId, string errorMsg) {

    }
  }
}

File: NCServer/src/ncserver/connection/HttpSessionBasedOnJson.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package ncserver.connection;

import com.sun.net.httpserver.Headers;
import java.util.HashMap;
import java.util.concurrent.atomic.AtomicReference;
import ncserver.utils.DataReader;
import org.json.JSONException;
import org.json.JSONObject;

/**
 *
 * @author Vladimir
 */
public class HttpSessionBasedOnJson extends HttpSession {

    @Override
    public boolean receiveData(String command, Headers headers, Object data, int size) {
        JsonMessageHandler handler = inJsonMessageHandlers.get(command);
        // Проверяем есть ли обработчик для сообщения такого типа
        if (handler == null) {
            onReceivingError("Internal error: unknown command");
            return false;
        }
        try {
            DataReader dataReader = new DataReader((byte[])data, 0, size);
            String json = "";
            AtomicReference<String> jsonRef = new AtomicReference<>(json);
            if (!dataReader.readString(jsonRef)) {
                onReceivingError("Json string is invalid");
                return false;
            }
            json = jsonRef.get();
            JSONObject jsonObject = json.length() > 1 ? new JSONObject(json) : new JSONObject("{}");
            // Обрабатываем сообщение - вызываем соответствующий обработчик                        
            return handler.call(headers, jsonObject);
        }
        catch (JSONException ex) {
            onReceivingError(ex.getMessage());
            return false;
        }
        catch (Exception exc) {
            String s = String.format("Internal error %s", exc.getMessage());
            onReceivingError(s);
            return false;
        }
    }

    public HttpSessionBasedOnJson(IConnectManager connManager, IConnectionInfo connection) {
        super(connManager, connection);
    }

    protected abstract class JsonMessageHandler {
        protected abstract boolean call(Headers headers, JSONObject jsonObject);
    }

    private final HashMap<Object, JsonMessageHandler> inJsonMessageHandlers = new HashMap<Object, JsonMessageHandler>();

    protected final void addMessageHandler(Object type, JsonMessageHandler messageHandlers) {
        inJsonMessageHandlers.put(type, messageHandlers);
    }

    @Override
    protected void onSessionClosing() {
        super.onSessionClosing();
        inJsonMessageHandlers.clear();
    }
}
